service/history/historyEngine.go

```diff
@@ -2013,16 +2018,53 @@ func (e *historyEngineImpl) DeleteWorkflowExecution(
 	if err != nil {
 		return err
 	}
-	defer func() { workflowContext.GetReleaseFn()(retError) }()
+	defer func() { weCtx.GetReleaseFn()(retError) }()
+
+	// Open and Close workflow executions are deleted differently.
+	// Open workflow execution is deleted by terminating with special flag `deleteAfterTerminate` set to true.
+	// This flag will be carried over with CloseExecutionTask and workflow will be deleted as the last step while processing the task.
+
+	// Close workflow execution is deleted using DeleteExecutionTask.
+
+	// DeleteWorkflowExecution is not replicated automatically. Workflow execution must be deleted separately in each cluster.
+	// Although running workflows in active cluster are terminated first and the termination is replicated.
+	// In passive cluster, workflow executions are just deleted in regardless of its status.
 
+	if weCtx.GetMutableState().IsWorkflowExecutionRunning() &&
+		namespaceEntry.ActiveInCluster(e.shard.GetClusterMetadata().GetCurrentClusterName()) {
+
+		// If workflow execution is running and in active cluster.
+		return api.UpdateWorkflowWithNew(
+			e.shard,
+			ctx,
+			weCtx,
+			func(workflowContext api.WorkflowContext) (*api.UpdateWorkflowAction, error) {
+				mutableState := workflowContext.GetMutableState()
+				eventBatchFirstEventID := mutableState.GetNextEventID()
+
+				return api.UpdateWorkflowWithoutWorkflowTask, workflow.TerminateWorkflow(
+					mutableState,
+					eventBatchFirstEventID,
+					"Delete workflow execution",
+					nil,
+					consts.IdentityHistoryService,
+					true,
+				)
+			},
+			nil)
+	}
+
+	// If workflow execution is closed or in passive cluster.
 	return e.workflowDeleteManager.AddDeleteWorkflowExecutionTask(
 		ctx,
-		nsID,
+		namespaceEntry.ID(),
 		commonpb.WorkflowExecution{
 			WorkflowId: request.GetWorkflowExecution().GetWorkflowId(),
 			RunId:      request.GetWorkflowExecution().GetRunId(),
 		},
-		workflowContext.GetMutableState(),
+		weCtx.GetMutableState(),
+		e.shard.GetQueueAckLevel(tasks.CategoryTransfer).TaskID,
+		e.shard.GetQueueAckLevel(tasks.CategoryVisibility).TaskID,
```

Comment:
Use cluster ack level and current cluster name? Then the delay is 30s. I can take care of the deprecation of cluster ack level later.